superset/connectors/sqla/models.py

```diff
@@ -2303,18 +2309,14 @@ def write_shadow_dataset(  # pylint: disable=too-many-locals
             # find referenced tables
             parsed = ParsedQuery(dataset.sql)
             referenced_tables = parsed.tables
-
-            # predicate for finding the referenced tables
-            predicate = or_(
-                *[
-                    and_(
-                        NewTable.schema == (table.schema or dataset.schema),
-                        NewTable.name == table.table,
-                    )
-                    for table in referenced_tables
-                ]
+            tables = load_or_create_tables(
+                session,
+                dataset.database_id,
+                dataset.schema,
+                referenced_tables,
+                conditional_quote,
+                engine,
             )
-            tables = session.query(NewTable).filter(predicate).all()
```

Comment:
We were only assigning tables that already exist. The `load_or_create_tables` function will create any tables that are referenced in the SQL but don't exist yet.